pages/projects/index.tsx

```diff
@@ -1,20 +1,68 @@
-import React from 'react';
-import {
-  // Search,
-  ProjectsList,
-  SharedLayout,
-} from '../../components';
-const Projects = () => {
+import React, { useState } from 'react';
+import { ProjectsList, SharedLayout } from '../../components';
+import { prisma } from '@/lib/prisma';
+import { InferGetServerSidePropsType } from 'next';
+import MultiSelectInput from '@/components/Form/MultiSelectInput/MultiSelectInput';
+import { useMemo } from 'react';
+import { RemoveTagFc, Tags } from '@/components/Tags';
+
+const Projects = ({
+  availableTags,
+}: InferGetServerSidePropsType<typeof getServerSideProps>) => {
+  const [selectedTags, setSelectedTags] = useState<string[]>([]);
+
+  const filteredAvailableTags = useMemo(() => {
+    return availableTags.filter((tag) => !selectedTags.includes(tag));
+  }, [availableTags, selectedTags]);
+
+  const onEnterClick = (value: string) => {
+    setSelectedTags((prev) => [...prev, value]);
+  };
+
+  const onDeleteTag: RemoveTagFc = (_event, index) => {
+    setSelectedTags((prev) => {
+      const newTags = [...prev];
+      newTags.splice(index, 1);
+      return newTags;
+    });
+  };
+
   return (
     <SharedLayout title="Projects" hideFooter>
       <div className="flex flex-col">
-        {/* <div className="sticky top-0 z-10 bg-yellow-400 backdrop-blur-3xl">
-          <Search />
-        </div> */}
-        <ProjectsList />
+        <div className="container m-auto flex max-w-screen-xl flex-col gap-y-4 md:px-5">
+          <div className="flex items-center gap-x-2">
+            <h4>Filter by tags:</h4>
+            <MultiSelectInput
+              onEnterClick={onEnterClick}
+              suggestions={filteredAvailableTags}
+              inputClassName="flex-1 w-full items-center rounded-md border-2 border-gray-200 bg-transparent p-1 outline-none focus:border-blue-600 focus:ring-blue-600"
+            />
+          </div>
+          <Tags
+            tags={selectedTags}
+            removeTagHandler={onDeleteTag}
+            className="gap-2"
+          />
+        </div>
+        <ProjectsList selectedTags={selectedTags} />
       </div>
     </SharedLayout>
   );
 };
 
 export default Projects;
+
+export async function getServerSideProps() {
+  const allProjects = await prisma.project.findMany();
+  const allTags = new Set<string>();
+  allProjects.forEach((project) =>
+    project.tags.forEach((tag) => allTags.add(tag))
+  );
+
+  return {
+    props: {
+      availableTags: Array.from(allTags),
+    },
+  };
+}
```

Comment:
I think fetching all projects to extract available tags at page request might slow page loading when there would be big amount of projects to fetch.

For now this solution is fine but we could think about adding some back-end logic that will store and track which tags are used in projects and how many times they appear? The tags data could be reached at specific endpoint. I could speed up loading.

What do you think about it? @rohitdasu @matmont

Reply:
i agree

Reply:
in future if we have more tags, it will be slow. need to think of the better solution. as of now it's fine though.

Reply:
yeah I agree, this would also help for the auto-completion when setting tags on projects.